documentation/src/main/asciidoc/user_guide/chapter-63-Streams.adoc

```diff
@@ -0,0 +1,416 @@
+method allows for distributed computations unlike if you manually retrieve these entries using
+link:https://docs.jboss.org/infinispan/{infinispanversion}/apidocs/org/infinispan/AdvancedCache.html#getAll-java.util.Set-[getAll]
+and you would have to process locally (pick the best method for the job).
+
+==== Segment based filtering
+
+NOTE: This is an advanced feature and should only be used with deep knowledge of Infinispan segment and hashing techniques.
+
+This option is only supported for replicated and distributed caches.  This allows the user to operate upon
+a subset of data at a time as determined by the
+link:https://docs.jboss.org/infinispan/{infinispanversion}/apidocs/org/infinispan/distribution/DistributionManager.html[DistributionManager].
+The segments can be filtered by invoking
+link:https://docs.jboss.org/infinispan/{infinispanversion}/apidocs/org/infinispan/CacheStream.html#filterKeySegments-java.util.Set-[filterKeySegments]
+method on the `CacheStream`.  This is applied after the key filter above but before any intermediate operations are performed.
+
+=== Local/Invalidation/Replication Cache
+
+A stream used with a local, invalidation or replication cache can be used just the same way you would use a stream on a
+regular collection. Infinispan handles all of the translations if necessary behind the scenes and works with all
+of the more interesting options (ie. storeAsBinary, compatibility mode, and a cache loader).
+
+An example of a simple stream task is as follows
+
+==== Example
+
+Below takes a given map and assuming the values are Strings it filters all of the entries that contain the
+word JBoss and then creates a new Map with only the results of the entries that contained JBoss as a value.
+[source,java]
+----
+Map<Object, String> jbossValues = cache.entrySet().stream()
+              .filter(e -> e.getValue().contains("Jboss"))
+              .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+----
+
+=== Distribution Cache
+
+This is where streams come into their stride.  When a stream operation is performed it will send the various intermediate and
+terminal operations to each node that has pertinent data.  This allows the resulting data to be sent back to the
+originating node and intermediate values are not required to be returned which results in much better performance.
+
+
+==== Rehash Aware
+
+Internally the data is segmented and each node only performs the operations upon the data it owns as a primary owner.  This allows
+for data to be processed evenly assuming segments are granular enough to provide for equal amounts of data on each node.
+
+When you are utilizing a distributed cache data this data can be quite volatile moving between nodes when a new node is added
+or an old one has left.  When this occurs data is shuffled between nodes.  This can cause issues with processing that data as it
+can see a subset or values a second time.
```

Comment:
I'm not sure saying "data can be quite volatile" is a good idea :)
I'd either give a concrete example of how processing data can be impacted by rehashing (distributed tasks?) or just mention that data is shuffled between nodes, and how distributed streams deal with it.